### PR TITLE
Modify EVP_CIPHER_is_a() and EVP_MD_is_a() to handle legacy methods too

### PR DESCRIPTION
**Pull request by levitte**
These functions would only handle provided methods, but there are
cases where the caller just passes along a received method without
knowing the underlying method tech, so might pass along a legacy
method.  We therefore need to have them handle this case as well so
they don't cause any unnecessary surprises.

### CODE REVIEW

File: test/namemap_internal_test.c [hunk 1]
```diff
@@ -108,12 +108,53 @@ static int test_cipherbyname(void)
...
+ */
+static int test_cipher_is_a(void)
+{
+    EVP_CIPHER *fetched = EVP_CIPHER_fetch(NULL, "AES-256-CCM", NULL);
+    int rv = 1;
+
+    if (!TEST_ptr_ne(fetched, NULL))
+        return 0;
+    if (!TEST_true(EVP_CIPHER_is_a(fetched, "id-aes256-CCM")))
+        rv = 0;
```
Comment (slontis) [inline test/namemap_internal_test.c]: may as well add a negative test here also.

File: test/namemap_internal_test.c [hunk 2]
```diff
@@ -108,12 +108,53 @@ static int test_cipherbyname(void)
     return 1;
 }
 
+/*
+ * Test that EVP_CIPHER_is_a() responds appropriately, even for cipherss that
```
Comment (slontis) [inline test/namemap_internal_test.c]: make it less snakey by removing a s

**Review by slontis — COMMENTED** [review]
I am not quite sure I understand why the legacy name cant just be put into the provider tables?

**Comment by levitte** [comment]
> I am not quite sure I understand why the legacy name cant just be put into the provider tables?

Ah, it seems I was a bit unclear.

The main point of this exercise is that legacy EVP_CIPHERs and EVP_MDs have an unset `name_id`, so it can't be used to check that it name that's tested has the same number.

Or did you mean that we would insert the legacy *NID* into the namemap? Yup, that was my first thought too, but that turned out to be more difficult, I ran into corner cases where names that weren't pre-populated didn't get the legacy NID, even though they were new aliases for pre-populated names, so evp_is_a ended up giving the wrong answer... and the code to do all that became quite ugly, so...

File: test/namemap_internal_test.c (hunk 1 above)
Comment (levitte) [reply test/namemap_internal_test.c]: Done

File: test/namemap_internal_test.c (hunk 2 above)
Comment (levitte) [reply test/namemap_internal_test.c]: Done

File: crypto/evp/evp_fetch.c [hunk 3]
```diff
@@ -444,11 +444,18 @@ const char *evp_first_name(OSSL_PROVIDER *prov, int name_id)
     return ossl_namemap_num2name(namemap, name_id, 0);
 }
 
-int evp_is_a(OSSL_PROVIDER *prov, int number, const char *name)
+int evp_is_a(OSSL_PROVIDER *prov, int number,
+             const char *legacy_name, const char *name)
 {
+    /*
+     * For a |prov| that is NULL, the library context will be NULL, which
+     * is fine, because that's the only place legacy stuff is relevant.
```
Comment (mattcaswell) [inline crypto/evp/evp_fetch.c]: I don't think this statement is true. I fully expect "legacy" ciphers/digests to be used in conjunction with non-default OPENSSL_CTXs. AFAIK there is nothing in the documentation suggesting you shouldn't be allowed to do that.

**Review by mattcaswell — CHANGES_REQUESTED** [review]

File: crypto/evp/evp_fetch.c (hunk 3 above)
Comment (levitte) [reply crypto/evp/evp_fetch.c]: Er...  so how do you associate legacy cipher/digests with a non-default OPENSSL_CTX?

Comment (levitte) [reply crypto/evp/evp_fetch.c]: I'm at a loss what to think or how to comment.  Ideas?

Comment (levitte) [reply crypto/evp/evp_fetch.c]: For everyone's information, @mattcaswell and I have a private chat where we concluded that this is not ideal, but ideal is hard to come by, so we're resorting to accepting this situation for legacy ciphers and just making sure it's documented.

Done

**Comment by levitte** [comment]
Ping

**Review by mattcaswell — APPROVED** [review]

**Comment by levitte** [comment]
Merged.

e4a1d0230016d090ba78bc7092384315f85b0e72 Modify EVP_CIPHER_is_a() and EVP_MD_is_a() to handle legacy methods too

**Pull request closed without merging**